Use JUnit 5 for JUnit test cases

kriegfrj opened the pull request:
Example of how JUnit 5 support might be added to the Bnd build.

This PR does the following:

* Adds library packages for JUnit Platform, and the Jupiter and Vintage engines to ext/central.mvn
* Adds macros to group them logically to ext/junit.bnd
* Converts `aQute.libg` to use the JUnit Platform.
  * `build.gradle` modified to cause Gradle to use the JUnit Platform rather than the JUnit 4 runner.
  * Converted `aQute.lib.io.IOTest` and `aQute.libg.uri.URIUtils` to use the Jupiter engine.
  * All other test classes remain as they were (JUnit 3 or 4 tests) and are executed by JUnit Platform using the Vintage engine.
* All other projects continue to use the JUnit 4 runner. To migrate them:
   1. Replace the ${junit} dependency in `-testpath` with ${junit-platform}
   2. Add the engines that you need to `-testpath` too (eg, ${junit-vintage} to run the old JUnit 3/4 tests, and ${junit-jupiter} if you add any Jupiter tests).

As a bonus, using the `@DisabledOnOs` annotation, I have managed to get rid of the test failures for the symlink tests on my machine.

Suggested future enhancements:
1. Get the bnd-gradle-plugin to add the necessary configuration when it detects junit-platform on the class path, so that a `build.gradle` file is not necessary.
2. Similar customisation (if appropriate) for the bnd-maven-m2e integration.

Comments welcomed/appreciated.

bjhargrave commented on aQute.libg/build.gradle:
```diff
@@ -0,0 +1,3 @@
+test {
+    useJUnitPlatform()
```
I don't like this. We do NOT want to put build.gradle files in every project. Anything we need to do for gradle for Junit 5 needs to be handled centrally in the root build.gradle file.

kriegfrj replied:
I didn't want to go big bang on all subprojects. I suppose it can still be added selectively from the root build.gradleI. The ideal solution, IMO, as I mentioned, is for the Gradle plugin to add it when it notices bnd.bnd has specified Junit Platform on the classpath.

kriegfrj replied:
I've changed it so that all projects are now using junit-platform.

bjhargrave commented on aQute.libg/bnd.bnd:
```diff
@@ -9,8 +9,9 @@
 -testpath: \
-	${junit},\
-	osgi.core;version=latest,\
+	${junit-platform},\
```
We want to have the `${junit}` macro represent the same jars for all projects to use for their `-testpath`. I don't want to have each project pick its own adventure here. This is why `${junit}` currently includes AssertJ. So I think we need a value for `${junit}` which all bundles will use.

Change this will require the `-buildpath` for `biz.aQute.junit` be changed to replace its use of `${junit}` with the current value of `${junit}`.

bjhargrave commented on cnf/ext/junit.bnd:
```diff
@@ -1,6 +1,20 @@
 junit: org.apache.servicemix.bundles.junit;version="[4.11,5)",\
- org.assertj.core;version="[3,4)"
+	org.assertj.core;version="[3,4)"
+junit-platform-version: 1.5.2
```
Not sure I like these version numbers in here. I think they should just be directly specified in central.mvn.

kriegfrj replied:
Your call. There are several modules in each group of dependencies that need to be kept at the same version, I thought using a macro for DRY was appropriate.

bjhargrave commented on aQute.libg/test/aQute/lib/io/IOTest.java:
```diff
@@ -297,8 +322,10 @@ public void testCreateSymlinkOrCopyWillDeleteOriginalLink() throws Exception {
-	public void testCreateDirectory_Symlink() throws Exception {
-		Path rootDirectory = Paths.get("generated/tmp/test/" + getName());
+	@Test
+	@DisabledOnOs(WINDOWS)
```
Why did you disable this on Windows? This test currently runs on Windows.

kriegfrj replied:
My bad. It didn't work on my box because I hadn't configured siblings to work properly. Easy to renable

bjhargrave commented on aQute.libg/test/aQute/lib/io/IOTest.java:
```diff
@@ -317,8 +344,10 @@ public void testCreateDirectory_Symlink() throws Exception {
-	public void testCreateDirectory_SymlinkMissingTarget() throws Exception {
-		Path rootDirectory = Paths.get("generated/tmp/test/" + getName());
+	@Test
+	@DisabledOnOs(WINDOWS)
```
Why did you disable this on Windows? This test currently runs on Windows.

kriegfrj replied:
Ditto.

bjhargrave commented on aQute.libg/test/aQute/lib/io/IOTest.java:
```diff
@@ -297,8 +322,10 @@ public void testCreateSymlinkOrCopyWillDeleteOriginalLink() throws Exception {
+	@Test
+	@DisabledOnOs(WINDOWS)
+	public void testCreateDirectory_Symlink(TestInfo info) throws Exception {
+		Path rootDirectory = Paths.get("generated/tmp/test/" + info.getTestMethod());
```
Also, this is wrong since `getTestMethod()` returns `Optional<Method>` and not a string. I think you need `info.getTestMethod().map(Method::getName).get()`

bjhargrave replied:
```java
	private String	testName;
	@BeforeEach
	public void before(TestInfo info) {
		testName = info.getTestMethod()
			.map(Method::getName)
			.get();
	}
```

kriegfrj replied:
I had falsely assumed that Method.toString() would be a friendly String of the method name but it's not that friendly at all. I've used getDisplayName() which prints as `testCreateDirectory_Symlink(TestInfo)` which should be good enough for a temp directory.

kriegfrj replied:
Changed it again - looked more closely at the second post there and saw that was better still & I've implemented that.

kriegfrj commented:
All changes implemented, ready for round 2.

kriegfrj commented:
Once I converted all subprojects to use junit-platform, the tests for `biz.aQute.bnd.gradle` started failing because it (along among all projects). I fixed it by adding `-testpath: ${junit}` to `bnd.bnd` (as I thought that would be the team's preference) but you could also possibly fix it by adding into `build.gradle`.

bjhargrave commented:
I will note, that I had to delete all my Junit launch configurations in Eclipse to properly launch Junit tests again. Since all my existing Junit launch configurations were using the Junit4 Runner, doing Run As > Junit Test for a project silently ignored the Junit 5 tests and only ran the Junit 4/3 tests!

Once I deleted all the Junit launch configurations, doing Run As > Junit Test for a project properly runs all tests. We will need to advise the other Bnd devs of this issue.

bjhargrave commented:
> Once I converted all subprojects to use junit-platform, the tests for `biz.aQute.bnd.gradle` started failing because it (alone among all projects) didn't have ${junit} on `-testpath` (in fact, it didn't have a `-testpath` at all). I fixed it by adding `-testpath: ${junit}` to `bnd.bnd` as I thought that would be the team's preference, but I also thought maybe the empty `-testpath` was a sign that we prefer to manage test dependencies through `build.gradle` instead for this project. This is also possible (in theory - haven't tried it in practice).

I assume that `test { useJUnitPlatform() }` in build.gradle requires Junit 5 jars on the test path which weren't otherwise there. The gradle project uses Spock which runs fine on the vintage platform. So your solution to set `-testpath` in the bnd file was correct.

kriegfrj commented:
> I will note, that I had to delete all my Junit launch configurations in Eclipse to properly launch Junit tests again. Since all my existing Junit launch configurations were using the Junit4 Runner, doing Run As > Junit Test for a project silently ignored the Junit 5 tests and only ran the Junit 4/3 tests!
> 
> Once I deleted all the Junit launch configurations, doing Run As > Junit Test for a project properly runs all tests. We will need to advise the other Bnd devs of this issue.

Yes, that's a trap - Junit 4 runner will silently ignore Jupiter tests. I don't know of an easy way to make that less silent.

Also might be worth pointing out you can set the default runner for the workspace. It's JUnit 5 for most workspaces already, but some may have set it to JUnit 4.
